src/data/padics/ring_homs.lean

```diff
@@ -0,0 +1,701 @@
+  push_cast,
+  rw [mul_right_comm, mul_assoc, ←this],
+  suffices rdcp : r.denom.coprime p,
+  { rw rdcp.gcd_eq_one, simp only [mul_one, cast_one, sub_self], },
+  apply coprime.symm,
+  apply (coprime_or_dvd_of_prime ‹_› _).resolve_right,
+  rw [← int.coe_nat_dvd, ← norm_int_lt_one_iff_dvd, not_lt],
+  apply ge_of_eq,
+  rw ← is_unit_iff,
+  exact is_unit_denom r h,
+end
+
+lemma norm_sub_mod_part (h : ∥(r : ℚ_[p])∥ ≤ 1) : ∥(⟨r,h⟩ - mod_part p r : ℤ_[p])∥ < 1 :=
+begin
+  let n := mod_part p r,
+  by_cases aux : (⟨r,h⟩ - n : ℤ_[p]) = 0,
+  { rw [aux, norm_zero], exact zero_lt_one, },
+  rw [norm_lt_one_iff_dvd, ← (is_unit_denom r h).dvd_mul_right],
+  suffices : ↑p ∣ r.num - n * r.denom,
+  { convert (int.cast_ring_hom ℤ_[p]).map_dvd this,
+    simp only [sub_mul, int.cast_coe_nat, ring_hom.eq_int_cast, int.cast_mul,
+      sub_left_inj, int.cast_sub],
+    apply subtype.coe_injective,
+    simp only [coe_mul, subtype.coe_mk, coe_coe],
+    rw_mod_cast @rat.mul_denom_eq_num r, refl },
+  dsimp [n, mod_part],
+  apply norm_sub_mod_part_aux r h,
+end
+
+lemma exists_mem_range_of_norm_rat_le_one (h : ∥(r : ℚ_[p])∥ ≤ 1) :
+  ∃ n : ℤ, 0 ≤ n ∧ n < p ∧ ∥(⟨r,h⟩ - n : ℤ_[p])∥ < 1 :=
+⟨mod_part p r, mod_part_nonneg _, mod_part_lt_p _, norm_sub_mod_part _ h⟩
+
+lemma zmod_congr_of_sub_mem_span_aux (n : ℕ) (x : ℤ_[p]) (a b : ℤ)
+  (ha : x - a ∈ (ideal.span {p ^ n} : ideal ℤ_[p]))
+  (hb : x - b ∈ (ideal.span {p ^ n} : ideal ℤ_[p])) :
+  (a : zmod (p ^ n)) = b :=
+begin
+  rw [ideal.mem_span_singleton] at ha hb,
+  rw [← sub_eq_zero, ← int.cast_sub,
+      zmod.int_coe_zmod_eq_zero_iff_dvd, int.coe_nat_pow],
+  rw [← dvd_neg, neg_sub] at ha,
+  have := dvd_add ha hb,
+  rwa [sub_eq_add_neg, sub_eq_add_neg, add_assoc, neg_add_cancel_left,
+      ← sub_eq_add_neg, ← int.cast_sub, pow_p_dvd_int_iff] at this,
+end
+
+lemma zmod_congr_of_sub_mem_span (n : ℕ) (x : ℤ_[p]) (a b : ℕ)
+  (ha : x - a ∈ (ideal.span {p ^ n} : ideal ℤ_[p]))
+  (hb : x - b ∈ (ideal.span {p ^ n} : ideal ℤ_[p])) :
+  (a : zmod (p ^ n)) = b :=
+zmod_congr_of_sub_mem_span_aux n x a b ha hb
+
+lemma zmod_congr_of_sub_mem_max_ideal (x : ℤ_[p]) (m n : ℕ)
+  (hm : x - m ∈ maximal_ideal ℤ_[p]) (hn : x - n ∈ maximal_ideal ℤ_[p]) :
+  (m : zmod p) = n :=
+begin
+  rw maximal_ideal_eq_span_p at hm hn,
+  have := zmod_congr_of_sub_mem_span_aux 1 x m n,
+  simp only [_root_.pow_one] at this,
+  specialize this hm hn,
+  apply_fun zmod.cast_hom (show p ∣ p ^ 1, by rw nat.pow_one) (zmod p) at this,
+  simpa only [ring_hom.map_int_cast],
+end
+
+variable (x : ℤ_[p])
+lemma exists_mem_range : ∃ n : ℕ, n < p ∧ (x - n ∈ maximal_ideal ℤ_[p]) :=
+begin
+  simp only [maximal_ideal_eq_span_p, ideal.mem_span_singleton, ← norm_lt_one_iff_dvd],
+  obtain ⟨r, hr⟩ := rat_dense (x : ℚ_[p]) zero_lt_one,
+  have H : ∥(r : ℚ_[p])∥ ≤ 1,
+  { rw norm_sub_rev at hr,
+    rw show (r : ℚ_[p]) = (r - x) + x, by ring,
+    apply le_trans (padic_norm_e.nonarchimedean _ _),
+    apply max_le (le_of_lt hr) x.2, },
```

Comment:
I am noticing a lot of proofs ending in `apply one_arg_function, apply a_term`. I thought the more mathlib-y style is to do just one call `exact one_arg_function a_term`. You can keep it as it is if you prefer the `apply` style.

Reply:
I've simplified a few, but especially when it forces a line break, I'm not sure there's much to be gained by condensing them. It's (slightly) easier to step through the proof this way.

Reply:
Yeah, I don't have any strong arguments either way, so only changing the obvious places is fine.